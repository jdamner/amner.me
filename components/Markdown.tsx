import ReactMarkdown from "react-markdown";

import { components } from "mdx-components";

export function Markdown({ children }) {
  return <ReactMarkdown components={components}>{children}</ReactMarkdown>;
}
