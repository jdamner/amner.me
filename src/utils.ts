/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}