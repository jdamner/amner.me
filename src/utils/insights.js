import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };