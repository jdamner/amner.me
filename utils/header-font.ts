import { Bebas_Neue as Font } from "next/font/google";
export const font = Font({
  weight: "400",
  subsets: ["latin", "latin-ext"],
});
