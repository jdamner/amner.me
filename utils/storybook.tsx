import { faker } from "@faker-js/faker";
import { PostLinkType } from "../types";
import React from "react";
import { parseMarkdown } from ".";
import matter from "gray-matter";

type MatterData = ReturnType<typeof matter>["data"];

faker.seed(123);

export const generateMdFile = <T extends MatterData>(attributes: T) =>
  parseMarkdown<T>(generateMD(attributes), "test.mdx");

export const generateMdContent = (): string =>
  `# ${faker.lorem.words(3)}\n` +
  Array.from(
    { length: 3 },
    () =>
      `## ${faker.lorem.words(2)}\n` +
      faker.lorem.paragraph() +
      "\n" +
      Array.from({ length: 3 }, () => {
        return `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`;
      }).join(""),
  ).join("");

export const generatePostLink = (): PostLinkType => ({
  title: faker.lorem.words(3),
  thumbnail: faker.image.url(),
  slug: faker.lorem.slug(),
});

export const generatePostLinks = (count: number): PostLinkType[] =>
  Array.from({ length: count }, () => generatePostLink());

export const generateMD = <T extends MatterData>(attributes: T) =>
  generateMDFrontMatter(attributes) + "\n" + generateMdContent();

export const generateMDFrontMatter = (attributes: MatterData) => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (size: number = 5) => (
  <>
    {Array.from({ length: size }).map((_, i) => (
      <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
    ))}
  </>
);
