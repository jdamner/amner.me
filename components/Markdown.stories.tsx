import type { Meta, StoryObj } from "@storybook/react";

import { generateMdContent } from "@/utils/storybook";
import { Markdown } from "./Markdown";

type Story = StoryObj<typeof Markdown>;

export default {
  title: "Components/Markdown",
  component: Markdown,
} as Meta;

export const Default: Story = {
  args: {
    children: generateMdContent(),
  },
};
