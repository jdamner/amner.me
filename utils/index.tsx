/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
