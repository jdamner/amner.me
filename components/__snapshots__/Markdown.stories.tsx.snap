// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Markdown Default smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <h1 id="subito-summisse-circumvenio"
      node="[object Object]"
  >
    subito summisse circumvenio
  </h1>
  <h2 id="curto-calamitas"
      node="[object Object]"
  >
    curto calamitas
  </h2>
  <p>
    Ipsa super super deprecator curo termes vomer cunctatio. Nisi denique apud cruciamentum cui conicio quis suspendo. Decens caries adulatio sub cubicularis.
  </p>
  <h3 id="paens-taceo"
      node="[object Object]"
  >
    paens taceo
  </h3>
  <p>
    Auctor decerno atque agnosco exercitationem summa exercitationem cursus. Coadunatio ulciscor deprimo surculus taedium perferendis contabesco suppono. Commodi sollers conturbo corrumpo calco canonicus.
  </p>
  <h3 id="claudeo-voveo"
      node="[object Object]"
  >
    claudeo voveo
  </h3>
  <p>
    Tener aliquam tergiversatio damno acsi curvo atavus depulso. Cursim aetas color valetudo. Tandem varietas suadeo vinco doloremque desipio.
  </p>
  <h3 id="delectus-quia"
      node="[object Object]"
  >
    delectus quia
  </h3>
  <p>
    Ancilla cognatus comes derideo cuppedia statua usitas. Casso amitto denuncio deputo voro candidus earum debilito placeat tantillus. Denique trepide thema.
  </p>
  <h2 id="pax-cimentarius"
      node="[object Object]"
  >
    pax cimentarius
  </h2>
  <p>
    Damnatio conicio absque cognomen confugo cur strenuus. Ustulo utique ceno dolore defluo sortitus censura occaecati. Quibusdam cerno spero ex turbo atque agnosco comburo tenax.
  </p>
  <h3 id="vallum-careo"
      node="[object Object]"
  >
    vallum careo
  </h3>
  <p>
    Autem depulso mollitia benevolentia. Eligendi vae artificiose. Verto supra vulgus acceptus demoror paens collum iusto.
  </p>
  <h3 id="dignissimos-articulus"
      node="[object Object]"
  >
    dignissimos articulus
  </h3>
  <p>
    Argumentum tendo sub amitto comis. Stultus verbera iste blanditiis creta conventus vespillo. Turba sum contabesco crudelis adimpleo campana cogo crastinus cubicularis.
  </p>
  <h3 id="inflammatio-sufficio"
      node="[object Object]"
  >
    inflammatio sufficio
  </h3>
  <p>
    Vulnero abeo constans reprehenderit tempus adsuesco optio. Stultus utique arcus. Cubitum tenax caput corona coniecto.
  </p>
  <h2 id="uterque-dolores"
      node="[object Object]"
  >
    uterque dolores
  </h2>
  <p>
    Sophismata ipsam amiculum. Aperte vado comminor degero. Aurum tyrannus claudeo ipsum toties ultra laboriosam creo.
  </p>
  <h3 id="spiritus-comes"
      node="[object Object]"
  >
    spiritus comes
  </h3>
  <p>
    Considero ustilo assumenda cura tricesimus adsum confido demulceo ipsam. Nesciunt nobis enim colligo abbas verto voveo. Ventosus conscendo beatus hic clarus deduco eius cimentarius.
  </p>
  <h3 id="veniam-clementia"
      node="[object Object]"
  >
    veniam clementia
  </h3>
  <p>
    Magnam causa sumo magni eaque timor consequuntur cruentus recusandae suus. Arto aspicio paulatim votum uredo consectetur vorago trepide ager. Cursus abscido bardus.
  </p>
  <h3 id="valens-defendo"
      node="[object Object]"
  >
    valens defendo
  </h3>
  <p>
    Infit aveho aliquid acsi clibanus totidem vestigium carbo modi adstringo. Aegre tardus demitto talis dolor administratio. Suggero cattus tumultus creber asperiores timor.
  </p>
</div>
`;
